src/util/measure_detector_to_reference.py

```diff
@@ -0,0 +1,103 @@
+import asyncio
+import hjson
+from ipaddress import IPv4Address
+import json
+import logging
+import numpy as np
+import os
+from time import sleep
+from timeit import main
+
+
+from src.board_builder.board_builder import BoardBuilder
+from src.common.structures.component_role_label import COMPONENT_ROLE_LABEL_DETECTOR, COMPONENT_ROLE_LABEL_POSE_SOLVER
+from src.controller.mct_controller import MCTController
+from src.controller.structures.mct_component_address import MCTComponentAddress
+
+input_filepath = "/home/adminpi5/Documents/MCSTrack/data/measure_detector_to_reference_config.json"
+
+with open(input_filepath, 'r') as file:
+    data = hjson.load(file)
+
+    startup_mode = data['startup_mode']
+    detectors = data['detectors']
+    pose_solvers = data['pose_solvers']
+    detector_params = data['detector_params']
+
+controller = MCTController(
+        serial_identifier="controller",
+        send_status_messages_to_logger=True)
+
+async def main():
+    logging.basicConfig(level=logging.INFO)
+
+    all_measured_transforms_by_detector = {}
+    board_builder = BoardBuilder()
+    detector_intrinsics = {}
+    ITERATIONS = 20
+
+    for detector in detectors:
+        controller.add_connection(MCTComponentAddress(
+                label=detector['label'],
+                role=COMPONENT_ROLE_LABEL_DETECTOR,
+                ip_address=detector['ip_address'],
+                port=detector['port']
+            ))
+        all_measured_transforms_by_detector[detector['label']] = []
+
+    controller.start_up()
+
+    while controller.is_transitioning():
+        await controller.update()
+
+    for i in range(ITERATIONS):
+
+        await controller.update()
+        detectors_and_their_frames = {}
+
+        for detector_label in controller.get_active_detector_labels():
+            board_builder.pose_solver.set_intrinsic_parameters(
+                detector_label, controller.get_live_detector_intrinsics(detector_label))
+
+            frame = controller.get_live_detector_frame(detector_label)
+            # Keep trying if it is a None frame, which happens on startup
+            while not frame:
+                await controller.update()
+                frame = controller.get_live_detector_frame(detector_label)
+
+            detectors_and_their_frames[detector_label] = frame.detected_marker_snapshots
+
+        board_builder.locate_reference_board(detectors_and_their_frames)
+    
+        for idx, detector_label in enumerate(controller.get_active_detector_labels()):
+            all_measured_transforms_by_detector[detector_label].append(\
+                board_builder.detector_poses[idx].object_to_reference_matrix.as_numpy_array())
```

Comment:
Interesting. I guess you're doing it this way because the current `pose_solver` is not working? Otherwise couldn't you just get the latest pose of the first (and only) target, which is the ChArUco board?

Reply:
That is a good point, I hadn't tried to implement it that way. I decided to go this route because I talked with David about his code to get the DetectorToReference and sort of tunnel-visioned on that. Do you want me to change it?

Reply:
No, if it works then leave it for now. There is currently a bug in the pose solver that needs fixing. Once it is fixed, then we can change this.